proposals/0023-typed-buffer-counters.md

```diff
@@ -24,39 +24,230 @@ explicit that they are separate resources.
+        int range, unsigned index, const char *name, unsigned
+        counterRegisterNo)`:
 
-### Multiple binding numbers on `resource.handlefrombinding`
+        Creates a resource with an explicit binding for the main buffer and an
+        explicit binding for its counter. The counter will be in the same space
+        as the main buffer.
+
+    *   `__createFromImplicitBindingWithCounter(unsigned orderId, unsigned
+        spaceNo, int range, unsigned index, const char *name, unsigned
+        counterRegisterNo)`:
+
+        Creates a resource with an implicit binding for the main buffer and an
+        explicit binding for its counter. The counter will be in the same space
+        as the main buffer.
+
+2.  **Sema Logic:** In `SemaHLSL.cpp`, the `initGlobalResourceDecl` function
+    will check if a resource type has a second field with a handle type that has
+    `[[hlsl::is_counter]]` attribute. If it does, it will emit a call to the
+    appropriate `...WithCounter` static method instead of the regular creation
+    methods.
+
+3.  **Handle Creation:** The `...With*Counter` methods, defined in
+    `clang/lib/Sema/HLSLBuiltinTypeDeclBuilder.cpp`, will initialize the two
+    handles using a combination of existing and new built-in functions defined
+    in `clang/include/clang/Basic/Builtins.td`.
+
+    *   The `__handle` (main data) will be initialized using the existing
+        `__builtin_hlsl_resource_handlefrombinding` or
+        `__builtin_hlsl_resource_handlefromimplicitbinding` built-ins.
+    *   To initialize the `__counter_handle`, new built-in functions will be
+        introduced: `__builtin_hlsl_resource_counterhandlefromimplicitbinding`
+        and `__builtin_hlsl_resource_counterhandlefrombinding`. These built-ins
+        will take the main resource handle (`__handle`) as an argument, along
+        with the counter's binding information.
+    *   During Clang's code generation, these new built-ins will be lowered to
+        their corresponding target-specific LLVM intrinsics for SPIR-V
+        (`llvm.spv.resource.counterhandle...`). For DirectX, since the counter
+        shares the same handle as the main resource, the built-in will be
+        replaced by a simple copy of the main resource handle. This approach
+        correctly models the relationship between the main resource and its
+        counter directly in the IR, as described in the "Proposed solution".
+
+### Array Handling
+
+For arrays of resources, the counter binding information is stored in the
+`HLSLCounterBindingAttr` of the array declaration itself. When Sema acts on the
+variable declaration, it will add an `HLSLCounterBindingAttr` with an implicit
+binding if the attribute does not already exist. When an array element is
+initialized with a call to a `__createHandle...` function during CodeGen, the
+appropriate `...With*Counter` version of the create function will be called for
+resources that have the `HLSLCounterBindingAttr`.
```

Comment:
```suggestion
For arrays of resources, the counter binding information is stored in the
`HLSLVkCounterBindingAttr` of the array declaration itself. When Sema acts on the
variable declaration, it will add an `HLSLVkCounterBindingAttr` with an implicit
binding if the attribute does not already exist. When an array element is
initialized with a call to a `__createHandle...` function during CodeGen, the
appropriate `...With*Counter` version of the create function will be called for
resources that have the `HLSLVkCounterBindingAttr`.
```

Reply:
I think it would be better to have all of the implicit binding order_ids stored in one place - on the `HLSLResourceBindingAttr`. So it would store both implicit order ID for the resource and for the counter (if the resources has a counter). Then if there is a `HLSLVkCounterBindingAttr` that clearly means explicit binding for the counter, which is the same behavior we have for `HLSLVkBinding`.